index.js

```diff
@@ -242,52 +209,129 @@ const parseArgs = (config = { __proto__: null }) => {
+      ArrayPrototypePush(
+        tokens,
+        { kind: 'option', name: longOption, rawName: `--${longOption}`,
+          index, value, inlineValue: true });
       continue;
     }
 
-    // Anything left is a positional
-    if (!allowPositionals) {
-      throw new ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL(arg);
+    ArrayPrototypePush(tokens, { kind: 'positional', index, value: arg });
+  }
+  return tokens;
+}
+
+const parseArgs = (config = { __proto__: null }) => {
+  const args = objectGetOwn(config, 'args') ?? getMainArgs();
+  const strict = objectGetOwn(config, 'strict') ?? true;
+  const allowPositionals = objectGetOwn(config, 'allowPositionals') ?? !strict;
+  const returnTokens = objectGetOwn(config, 'tokens') ?? false;
+  const options = objectGetOwn(config, 'options') ?? { __proto__: null };
+  // Bundle these up for passing to strict-mode checks.
+  const parseConfig = { args, strict, options, allowPositionals };
+
+  // Validate input configuration.
+  validateArray(args, 'args');
+  validateBoolean(strict, 'strict');
+  validateBoolean(allowPositionals, 'allowPositionals');
+  validateBoolean(returnTokens, 'tokens');
+  validateObject(options, 'options');
+  ArrayPrototypeForEach(
+    ObjectEntries(options),
+    ({ 0: longOption, 1: optionConfig }) => {
+      validateObject(optionConfig, `options.${longOption}`);
+
+      // type is required
+      validateUnion(objectGetOwn(optionConfig, 'type'), `options.${longOption}.type`, ['string', 'boolean']);
+
+      if (ObjectHasOwn(optionConfig, 'short')) {
+        const shortOption = optionConfig.short;
+        validateString(shortOption, `options.${longOption}.short`);
+        if (shortOption.length !== 1) {
+          throw new ERR_INVALID_ARG_VALUE(
+            `options.${longOption}.short`,
+            shortOption,
+            'must be a single character'
+          );
+        }
+      }
+
+      if (ObjectHasOwn(optionConfig, 'multiple')) {
+        validateBoolean(optionConfig.multiple, `options.${longOption}.multiple`);
+      }
     }
+  );
+
+  // Phase 1: identify tokens
+  const tokens = argsToTokens(args, options);
 
-    ArrayPrototypePush(result.positionals, arg);
+  // Phase 2: process tokens into parsed option values and positionals
+  const result = {
+    values: { __proto__: null },
+    positionals: [],
+  };
+  if (returnTokens) {
+    result.tokens = tokens;
   }
+  ArrayPrototypeForEach(tokens, (token) => {
+    switch (token.kind) {
+      case 'option-terminator':
+        break;
+      case 'positional':
+        if (!allowPositionals) {
+          throw new ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL(token.value);
+        }
+        ArrayPrototypePush(result.positionals, token.value);
+        break;
+      case 'option':
+        if (strict) {
+          checkOptionUsage(parseConfig, token);
+          checkOptionLikeValue(token);
+        }
+        storeOption(token.name, token.value, options, result.values);
+        break;
+    }
+  });
```

Comment:
```suggestion
    if (token.kind === 'positional') {
        if (!allowPositionals) {
          throw new ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL(token.value);
        }
        ArrayPrototypePush(result.positionals, token.value);
      } if (token.kind === 'option') {
        if (strict) {
          checkOptionUsage(parseConfig, token);
          checkOptionLikeValue(token);
        }
        storeOption(token.name, token.value, options, result.values);
    }
  });
```

as always, avoiding a switch is much shorter and less bug-prone :-)

Reply:
I sometimes agree that if/else better, but I like the switch statement for recognising all of a range of options in a visible way.

Reply:
since there's no `default`, `option-terminator` doesn't do anything, so even in the switch it should be removed (or, a default behavior added)

Reply:
Hmm, without static code analysis the switch statement covering the enumerated values is not benefiting future maintenance as it does in some other environments. I'll have a look at what JSDoc and ESLint might offer in combination before deciding whether to change.

Reply:
Switch statements aren’t forced to be exhaustive, so in JS this benefit only ever exists if the default path throws.

Pattern matching may be helpful here, but it’s still early.

Reply:
Ah yes, I saw a couple of patterns using throw or return to trigger checks or exceptions. There is an eslint check available in TypeScript: `switch-exhaustiveness-check`. 

I didn't use `switch` in any of the examples I wrote later, it was only this first one where it seemed a good fit. I'll refactor.